File: read_symbols.py
"""
read_symbols.py  —  Python equivalent of read_symbols.cpp
Reads and prints .rays/symbols.msgpack
Usage: python read_symbols.py <project_root>
"""
import sys
from pathlib import Path
import msgpack

def main():
    test_path = sys.argv[1] if len(sys.argv) > 1 else \
        "/Users/samreedhbhuyan/Desktop/Win_C/RAYS/YOLO_model_everything/RAYS_V5/trial_codebases/alphafold"

    msgpack_file = Path(test_path) / ".rays" / "symbols.msgpack"

    if not msgpack_file.exists():
        print(f"Error: {msgpack_file} does not exist!")
        print("Run 'python rays_generator.py' first to generate it.")
        sys.exit(1)

    with open(msgpack_file, "rb") as f:
        symbol_registry = msgpack.unpack(f, raw=False)

    print("==================================================")
    print(f"Total symbols found: {len(symbol_registry)}")
    print("==================================================\n")

    for record in symbol_registry:
        print(f"Symbol: {record['symbol_name']}")
        print(f"  Type:       {record['symbol_type']}")
        print(f"  File:       {record['file_path']}")
        print(f"  Location:   Line {record['start_line']}-{record['end_line']}")
        print(f"  Byte Range: {record['start_byte']}-{record['end_byte']}")
        print(f"  Visibility: {record['visibility']}")
        if record.get('parent_symbol'):
            print(f"  Parent:     {record['parent_symbol']}")
        print("--------------------------------------------------")

    print(f"\nTotal: {len(symbol_registry)} symbols")

if __name__ == "__main__":
    main()

File: read_symbols.cpp
#include "rays_generator.cpp"
#include <iostream>
#include <fstream>

int main() {
    std::string test_path = "/Users/samreedhbhuyan/Desktop/Win_C/RAYS/YOLO_model_everything/RAYS_V5/trial_codebases/alphafold";
    
    fs::path msgpack_file = fs::path(test_path) / ".rays" / "symbols.msgpack";
    
    if (!fs::exists(msgpack_file)) {
        std::cout << "Error: " << msgpack_file << " does not exist!" << std::endl;
        std::cout << "Run './test_rays' first to generate it." << std::endl;
        return 1;
    }
    
    std::ifstream ifs(msgpack_file, std::ios::binary);
    
    std::string buffer_str((std::istreambuf_iterator<char>(ifs)), std::istreambuf_iterator<char>());
    
    msgpack::object_handle oh = msgpack::unpack(buffer_str.data(), buffer_str.size());
    
    msgpack::object obj = oh.get();
    
    std::vector<SymbolRecord> symbol_registry;
    obj.convert(symbol_registry);
    
    std::cout << "==================================================" << std::endl;
    std::cout << "Total symbols found: " << symbol_registry.size() << std::endl;
    std::cout << "==================================================" << std::endl << std::endl;
    
    for (const auto& record : symbol_registry) {
        std::cout << "Symbol: " << record.symbol_name << std::endl;
        std::cout << "  Type: " << record.symbol_type << std::endl;
        std::cout << "  File: " << record.file_path << std::endl;
        std::cout << "  Location: Line " << record.start_line << "-" << record.end_line << std::endl;
        std::cout << "  Byte Range: " << record.start_byte << "-" << record.end_byte << std::endl;
        std::cout << "  Visibility: " << record.visibility << std::endl;
        
        if (!record.parent_symbol.empty()) {
            std::cout << "  Parent: " << record.parent_symbol << std::endl;
        }
        
        std::cout << "--------------------------------------------------" << std::endl;
    }
    
    std::cout << std::endl << "Total: " << symbol_registry.size() << " symbols" << std::endl;
    
    return 0;
}

